p2panda-rs/src/document/document.rs

```diff
@@ -0,0 +1,428 @@
+        // NB. This validation is quite excessive as it's normally not possible to get to this
+        // point while having broken some of these basic data restraints.
+
+        // Validate each operation in this document.
+        self.iter()?.try_for_each(|op| {
+            // If this is a delete operation check there are no fields.
+            if op.is_delete() {
+                if op.fields().is_none() {
+                    return Ok(())
+                } else {
+                    return Err(DocumentError::ValidationError("DELETE operations should not contain any fields".to_string()))
+                };
+            };
+            // Validate each update operation against the document schema.
+            match self.schema().validate_operation_fields(&op.fields().unwrap()) {
+                Ok(_) => Ok(()),
+                Err(_) => Err(DocumentError::ValidationError(
+                    "All CREATE and UPDATE operations in document must follow the schema description".to_string(),
+                )),
+            }
+        })?;
+
+        Ok(())
+    }
+}
+
+/// A struct for building documents.
+#[derive(Debug)]
+pub struct DocumentBuilder {
+    /// An unsorted collection of operations which are associated with a particular document id.
+    operations: Vec<OperationWithMeta>,
+}
+
+impl DocumentBuilder {
+    /// Instantiate a new DocumentBuilder with a collection of operations.
+    pub fn new(mut operations: Vec<OperationWithMeta>) -> Self {
+        // Sort operations alphabetically by their hash id.
+        // This is important for graph building and sorting, operations must be added
+        // to a IncrementalTopo graph in the correct (alphabetical) order to assure consistent
+        // traversal.
+        operations.sort_by(|a, b| a.operation_id().as_str().cmp(b.operation_id().as_str()));
+        Self { operations }
+    }
+
+    /// Get all operations for this document.
+    pub fn operations(&self) -> Vec<OperationWithMeta> {
+        self.operations.clone()
+    }
+
+    /// Get an iterator over all operations in this document.
+    pub fn operations_iter(&self) -> std::vec::IntoIter<OperationWithMeta> {
+        self.operations.clone().into_iter()
+    }
+
+    /// Build the document.
+    pub fn build(self) -> Result<Document, DocumentBuilderError> {
+        // find create message
+
+        let collect_create_operation: Vec<OperationWithMeta> =
+            self.operations_iter().filter(|op| op.is_create()).collect();
+
+        if collect_create_operation.len() > 1 {
+            return Err(DocumentBuilderError::MoreThanOneCreateOperation);
+        } else if collect_create_operation.is_empty() {
+            return Err(DocumentBuilderError::NoCreateOperation);
+        }
+
+        let create_operation = collect_create_operation.get(0).unwrap(); // unwrap as we know there is one item
+
+        // Validate the provided schema
+        Schema::new(&create_operation.schema(), DOCUMENT_SCHEMA)?;
```

Comment:
This is checking the validity of `DOCUMENT_SCHEMA` right now, there's no check as to whether `DOCUMENT_SCHEMA` belongs to the hash given in the first argument. I've been thinking we should maybe become clearer in what we mean we say 'validates' for example. Maybe change the comment to 

> Validates the provided schema's CDDL definition

Reply:
Yes, good call!